process/smartContract/builtInFunctions/keyValueStorage.go

```diff
@@ -36,54 +36,62 @@ func NewSaveKeyValueStorageFunc(
 func (k *saveKeyValueStorage) ProcessBuiltinFunction(
 	_, acntDst state.UserAccountHandler,
 	input *vmcommon.ContractCallInput,
-) (*big.Int, uint64, error) {
+) (*vmcommon.VMOutput, error) {
 	if input == nil {
-		return nil, 0, process.ErrNilVmInput
+		return nil, process.ErrNilVmInput
 	}
 	if len(input.Arguments) != 2 {
-		return nil, input.GasProvided, process.ErrInvalidArguments
+		return nil, process.ErrInvalidArguments
 	}
+	value := input.Arguments[1]
+	length := uint64(len(value))
+	maxGasToUse := k.funcGasCost + length*(k.gasConfig.DataCopyPerByte+k.gasConfig.StorePerByte)
+	if input.GasProvided < maxGasToUse {
+		return nil, process.ErrNotEnoughGas
+	}
+
 	if check.IfNil(acntDst) {
-		return nil, input.GasProvided, process.ErrNilSCDestAccount
+		// cross-shard call, in sender shard only the gas is taken out
+		return &vmcommon.VMOutput{ReturnCode: vmcommon.Ok}, nil
```

Comment:
It is not possible to call this cross-shard. as the calledAddress has to be equal with the acntDST. So it is only in shard. But instead of returning a stupid error for a bad call, I would just consume all the gas here.

Reply:
returned wvmcommon user error